src/cliffordsynthesis/Tableau.cpp

```diff
@@ -492,4 +492,107 @@ bool Tableau::isIdentityTableau() const {
   }
   return true;
 }
+
+Tableau Tableau::applyMapping(const std::vector<std::vector<bool>>* p) {
+  Tableau mappedTableau = Tableau(nQubits, hasDestabilizers());
+  for (size_t i = 0; i < mappedTableau.getTableauSize(); i++) {
+    for (unsigned char& j : mappedTableau.tableau[i]) {
+      j = 0;
+    }
+  }
+  for (size_t i = 0; i < p->size(); i++) {
+    for (size_t j = 0; j < (*p)[i].size(); j++) {
+      // apply mapping from column i to j if p is set
+      if ((*p)[i][j]) {
+        // in every row swap x entry and z entry
+        for (size_t n = 0; n < mappedTableau.getTableauSize(); n++) {
+          mappedTableau.tableau[n][j] = tableau[n][i];
+          mappedTableau.tableau[n][j + mappedTableau.nQubits] =
+              tableau[n][i + mappedTableau.nQubits];
+        }
+      }
+    }
+  }
+  // copy r column without changes
+  for (size_t i = 0; i < tableau.size(); i++) {
+    mappedTableau.tableau[i][2 * nQubits] = tableau[i][2 * nQubits];
+  }
+  return mappedTableau;
+}
+
+// number of Qubits is passed because nQubits is not set in result Tableau of
+// synthesis
+Tableau Tableau::reverseMappingOnRows(const std::vector<std::vector<bool>>* p,
+                                      const size_t nq) {
+  auto mappedTableau = Tableau(nq, true);
+  mappedTableau.tableau = tableau;
+  for (size_t i = 0; i < p->size(); i++) {
+    for (size_t j = 0; j < (*p)[i].size(); j++) {
+      // apply mapping from row i to j if p is set
+      if ((*p)[i][j]) {
+        mappedTableau.tableau[i] = tableau[j];
+        mappedTableau.tableau[i + nq] = tableau[j + nq];
+      }
+    }
+  }
+  return mappedTableau;
+}
+
+// in place Gauss Elimination
+void Tableau::gaussianEliminationGF2() {
```

Comment:
Minor comment: since you compute the reduced row echelon form of the matrix and use it's uniqueness to compare tableaus, calling this method something like `rref` would make it clearer what is actually computed.